rust/src/io/object_store.rs

```diff
@@ -52,16 +57,43 @@ impl std::fmt::Display for ObjectStore {
     }
 }
 
+#[derive(Debug)]
+struct AwsCredentialAdapter {
+    pub inner: DefaultCredentialsChain,
+}
+
+#[async_trait]
+impl CredentialProvider for AwsCredentialAdapter {
```

Comment:
Does this pickup credentials from EC2 IAM roles?

Ideally we'd support the following modes (forgot the right order):
1. Credentials as envvars
2. ~/.aws/credentials
3. IAM role on EC2 instance

Additionally, if the user has multiple profiles setup, then AWS_PROFILE envvar should also affect which sets of credentials are being used.

Reply:
I think that is the behavior for DefaultProviderChain, code here: https://github.com/awslabs/aws-sdk-rust/blob/d1962fd42d376c5c695877041d2a269fd1fa50cd/sdk/aws-config/src/default_provider/credentials.rs#L189-L193

The profile provider in that chain uses `AWS_PROFILE` envvar for selecting profile. (After many layers of abstractions, this is the code used for selecting profile name: https://github.com/awslabs/aws-sdk-rust/blob/d1962fd42d376c5c695877041d2a269fd1fa50cd/sdk/aws-config/src/profile/parser/source.rs#L59-L63)

Reply:
That said, let me test this.